tests/extended/test_trainer_ext.py

```diff
@@ -45,6 +45,9 @@
 set_seed(42)
 MARIAN_MODEL = "sshleifer/student_marian_en_ro_6_1"
 MBART_TINY = "sshleifer/tiny-mbart"
+T5_TINY = "patrickvonplaten/t5-tiny-random"
+# using T5_TINY as it's fp32 model, MBART_TINY which is fp16 fails under 0 gpus with various
+# Half-related issues, e.g. RuntimeError: "LayerNormKernelImpl" not implemented for 'Half'
```

Comment:
This makes me think we should check if there is a GPU available before auto-detecting the dtype and forcing the default dtype to float32 if there are none. Otherwise, the whole inference API on CPU will start failing.

So we should actually leave the model as it was for a good test of that!

Reply:
Isn't this a bug? Why do we want fp16 models loaded as fp32? 

Pytorch doesn't have all ops implemented on CPU for Half. But I don't think we should automatically compensate for that, since it'd prevent situations where it works just fine. If the model.half doesn't work under CPU it's not a bug, but a pytorch limitations. Users can request the missing features to be implemented.

> So we should actually leave the model as it was for a good test of that!

You meant, leaving the breaking test as is, rather than working around it - and finding a fix instead.

Then I would rather make a special test which tests this thing, otherwise this test caught the problem absolutely accidentally.

Reply:
Moreover, while it's a very sensible suggestion, checking if GPU is available won't guarantee the user using GPU.

Reply:
> Why do we want fp16 models loaded as fp32?

Because FP16 does not work on CPU and if all the models saved in FP16 to make downloading them easier (since they are half the size) suddenly fail for all our users, we have a very big problem.

> Users can switch the model to fp32 if they need to.

It's not because the breaking change has a workaround that it's still a breaking change. The end user code stops working and this is something we do not want.

Thinking more about this, I don't think the model config is a good option in the end, since a user will upload a model on the Hub with a FP16 in the config and that model won't work for the inference API or all users on CPU. It should really be completely opt-in.

cc @LysandreJik and @patrickvonplaten

Reply:
I hear you about breaking behavior is still a breaking even if it's doing the right thing. 

So how about this:

1. plan to automate this in v5 and meanwhile do the right thing only if it's triggered manually via `from_pretrained()` / `from_config()` new `torch_dtype` arg
2. I propose we still save `torch_dtype` in the config in `save_pretained`, so that the newly added models start populating this data, but we won't use it for now. The added bonus is that the hub could immediately tell the visitor the dtype of this model - much cheaper than loading the model and checking its weights.

Reply:
For half-precision inference on CPU pytorch is transitioning to bfloat16 
https://github.com/pytorch/pytorch/issues/23509

Reply:
I agree with defaulting to FP32 when FP16 is required but the platform used doesn't support it, like CPU. I'm down to have the `torch_dtype` be populated and eventually used if on a platform that supports it.

And regarding the automation in v5 let's visit this once we switch.

Reply:
That's the thing:
1. we don't know inside modeling_tools if the platform supports it.
2. we don't know what the eventual use will be even if somehow we could check the env. a model can be loaded under cpu and then switched to gpu, and vice versa.

Reply:
Then if I understand correctly, I'm fine with:
- Support `torch_dtype` in the config
- Only use it to update the weights if triggered manually with `from_pretrained` or `from_config`.

Switching automatically in v5 will require revisiting, as right now I'm not eager on having some models fail by default on a certain number of platforms

Reply:
OK, recoded to add `torch_dtype` arg to `from_pretrained` and `from_config`

Also added `torch_dtype_auto_detect` to `from_pretrained`